File: src/crush/interfaces/api/apercu.py
# Copyright (C) 2026 Maxime Song

"""« Dans quel état est mon assistant, là, maintenant ? » — en un seul écran.

CE QUE LES AUTRES PAGES NE FONT PAS

L'Écosystème est une liste de contrôle : maillon par maillon, son état et son
remède. On l'ouvre quand quelque chose cloche. Le Cerveau montre la FORME de
l'assistant, pas son état. Le Pilotage montre ce qui demande une décision.

Aucune ne répond à la question qu'on se pose en s'asseyant : est-ce que tout
tourne, qu'est-ce qui attend, combien ça coûte, et quand la prochaine passe. Ces
réponses existaient, éparpillées sur cinq pages et six appels réseau.

UN SEUL APPEL

Tout est agrégé ici plutôt que composé côté navigateur. Six requêtes depuis un
téléphone en 4G, c'est six aller-retours et un écran qui se remplit par morceaux
pendant deux secondes — or c'est précisément la page qu'on ouvre pour un coup
d'œil rapide.

AUCUNE MESURE INVENTÉE

Même règle que l'Écosystème : ce qui ne peut pas être constaté est rendu `null`,
jamais approché. Un aperçu qui invente une valeur plausible est pire qu'un champ
vide, parce qu'on cesse de vérifier ailleurs.
"""

from __future__ import annotations

from datetime import datetime
from typing import Any

from fastapi import APIRouter, Request

from crush.interfaces.api.ecosysteme import canal_actif, ecosysteme
from crush.kernel.paths import MEMORY_DATA_DIR, SAUVEGARDES_DIR
from crush.kernel.remote_agents import registry
from crush.kernel.settings import settings

router = APIRouter()

_CANAUX = ("telegram", "discord", "signal", "slack", "whatsapp")


def _age_heures(horodatage: float) -> float:
    return (datetime.now().timestamp() - horodatage) / 3600


def _lisible(heures: float | None) -> str:
    if heures is None:
        return "inconnu"
    if heures < 1:
        return "il y a moins d'une heure"
    if heures < 48:
        return f"il y a {int(heures)} h"
    return f"il y a {int(heures / 24)} jours"


def _cerveau() -> dict[str, Any]:
    backend = settings.api_backend
    modele = {
        "gemini": settings.gemini_model,
        "anthropic": settings.anthropic_model,
        "openai": settings.openai_model,
        "mistral": settings.mistral_model,
    }.get(backend, backend)
    return {
        "backend": backend,
        "modele": modele,
        "reflexion": bool(settings.reflection_enabled),
        "voix": settings.tts_provider,
    }


def _memoire(request: Request) -> dict[str, Any]:
    base = MEMORY_DATA_DIR / "crush_memory.db"
    kernel = getattr(request.app.state, "memory_kernel", None)

    faits = None
    if kernel is not None:
        try:
            from crush.providers.memory.schemas import FactStatus

            faits = kernel.count_facts(FactStatus.ACTIVE)
        except Exception:  # noqa: BLE001 — un compteur absent vaut mieux qu'un faux
            faits = None

    archives = sorted(SAUVEGARDES_DIR.glob("memoire-*.tar.gz")) if SAUVEGARDES_DIR.is_dir() else []
    age = _age_heures(archives[-1].stat().st_mtime) if archives else None

    return {
        "faits": faits,
        "octets": base.stat().st_size if base.exists() else None,
        "archives": len(archives),
        "sauvegarde_age_h": round(age, 1) if age is not None else None,
        "sauvegarde_lisible": _lisible(age),
        # Plus de 36 h : la passe quotidienne ne tourne manifestement plus. Le
        # seuil laisse passer une nuit d'extinction sans crier au loup.
        "sauvegarde_inquiete": age is not None and age > 36,
    }


def _cout(request: Request) -> dict[str, Any]:
    """Le coût réel, ou `null`. Jamais une estimation présentée comme un relevé."""
    tracker = getattr(request.app.state, "tracker", None)
    if tracker is None:
        return {"aujourd_hui": None, "mois": None, "serie": [], "devise": "EUR"}
    try:
        jours = tracker.get_daily_totals(days=7)
        mois = tracker.get_monthly_totals()
    except Exception:  # noqa: BLE001
        return {"aujourd_hui": None, "mois": None, "serie": [], "devise": "EUR"}

    # `cost_usd` est la cle reelle du tracker, et la devise est le DOLLAR. La
    # premiere version essayait une liste de cles plausibles et retombait sur
    # 0.0 : elle affichait donc « 0,00 € » -- un montant faux, dans la mauvaise
    # monnaie, presente comme un releve. Une cle absente rend `None`.
    def _valeur(entree: object) -> float | None:
        if isinstance(entree, dict) and isinstance(entree.get("cost_usd"), (int, float)):
            return float(entree["cost_usd"])
        return None

    valeurs = [_valeur(j) for j in (jours or [])]
    serie = [round(v, 4) for v in valeurs if v is not None]
    total_mois = _valeur(mois)

    # Rien n'a jamais ete suivi : « 0 » laisserait croire a une depense nulle
    # alors que la mesure n'a simplement pas commence.
    suivi_depuis = (mois or {}).get("tracked_since") if isinstance(mois, dict) else None
    if suivi_depuis is None:
        return {"aujourd_hui": None, "mois": None, "serie": [], "devise": "USD"}

    return {
        "aujourd_hui": serie[-1] if serie else None,
        "mois": round(total_mois, 2) if total_mois is not None else None,
        "serie": serie,
        "devise": "USD",
        "suivi_depuis": suivi_depuis,
    }


def _initiatives(request: Request) -> dict[str, Any]:
    store = getattr(request.app.state, "initiative_store", None)
    if store is None:
        try:
            from crush.engine.proactive.store import InitiativeStore

            store = InitiativeStore()
        except Exception:  # noqa: BLE001
            return {"en_attente": None, "haute": 0, "liste": []}
    try:
        en_attente = store.load_pending_all(days=7)
    except Exception:  # noqa: BLE001
        return {"en_attente": None, "haute": 0, "liste": []}

    def _priorite(i: object) -> str:
        return str(getattr(getattr(i, "priority", ""), "value", getattr(i, "priority", "")))

    return {
        "en_attente": len(en_attente),
        "haute": sum(1 for i in en_attente if _priorite(i) == "high"),
        "liste": [
            {
                "id": getattr(i, "id", ""),
                "titre": getattr(i, "title", ""),
                "priorite": _priorite(i),
                # Le POURQUOI, en une ligne. Trancher sans lui obligeait a
                # ouvrir le Command Center pour chaque item -- et c'est
                # precisement cette corvee qui a laisse la file monter a 25.
                "pourquoi": str(getattr(i, "reasoning", "") or getattr(i, "context", ""))[:160],
                "type": str(getattr(getattr(i, "type", ""), "value", getattr(i, "type", ""))),
                "decision": str(
                    getattr(getattr(i, "execution_mode", ""), "value", "")
                ).lower()
                == "validate",
            }
            for i in en_attente[:12]
        ],
    }


def _relie() -> list[dict[str, Any]]:
    """Ce qui est effectivement branché, et ce qui dort. La distinction compte :
    un canal éteint volontairement n'est pas une panne."""
    liens: list[dict[str, Any]] = []
    for canal in _CANAUX:
        liens.append(
            {
                "nom": canal.capitalize(),
                "actif": canal_actif(canal),
                "famille": "canal",
            }
        )
    for agent in registry.list_agents():
        liens.append(
            {
                "nom": agent.name,
                "actif": True,
                "famille": "machine",
                "detail": f"{agent.platform} · {len(agent.actions)} action(s)",
            }
        )
    liens.append(
        {
            "nom": "Coffre Obsidian",
            "actif": (MEMORY_DATA_DIR / "mirror" / "boite-de-reception.md").exists(),
            "famille": "memoire",
        }
    )
    return liens


@router.get("/api/apercu")
async def lire_l_apercu(request: Request) -> dict[str, Any]:
    """Tout ce qu'il faut pour un coup d'œil, en une requête."""
    presence = getattr(request.app.state, "presence", None)
    etat_presence: dict[str, Any] = {"resume": None, "joignable": None, "au_poste": False}
    if presence is not None:
        try:
            e = await presence.etat()
            etat_presence = {
                "resume": e.resume(),
                "joignable": e.joignable,
                "au_poste": e.au_poste,
                "erreur": e.erreur,
            }
        except Exception:  # noqa: BLE001 — l'aperçu ne tombe pas pour un champ
            pass

    scheduler = getattr(request.app.state, "scheduler", None)
    boucles: list[dict[str, Any]] = []
    if scheduler is not None:
        try:
            boucles = [
                {
                    "nom": b.get("name", ""),
                    "quand": b.get("next_run"),
                    "cadence": b.get("interval", ""),
                    "detail": b.get("description", ""),
                }
                for b in scheduler.status()
            ]
        except Exception:  # noqa: BLE001
            boucles = []

    # On réutilise l'Écosystème plutôt que de recompter : deux comptages qui
    # divergent enverraient chercher une panne sur la mauvaise page.
    resume_eco: dict[str, Any] = {}
    try:
        vue = await ecosysteme(request)
        resume_eco = dict(vue.get("resume") or {})
        resume_eco["a_regarder"] = [
            {"nom": m["nom"], "etat": m["etat"], "detail": m["detail"], "remede": m["remede"]}
            for m in vue.get("maillons", [])
            if m["etat"] != "ok"
        ][:5]
    except Exception:  # noqa: BLE001
        resume_eco = {}

    return {
        "horodatage": datetime.now().isoformat(timespec="seconds"),
        "assistant": settings.assistant_name or "Crush",
        "cerveau": _cerveau(),
        "memoire": _memoire(request),
        "cout": _cout(request),
        "initiatives": _initiatives(request),
        "presence": etat_presence,
        "relie": _relie(),
        "boucles": boucles,
        "ecosysteme": resume_eco,
        "silence": {
            "plage": settings.push_heures_silence,
            "urgent_passe": bool(settings.push_silence_laisse_passer_urgent),
        },
    }

File: src/crush/interfaces/ui/static/apercu.js
/* apercu.js — « dans quel état est mon assistant, là, maintenant ? »
 *
 * L'Écosystème est une liste de contrôle qu'on ouvre quand ça cloche. Le Graphe
 * montre la forme de l'assistant, pas son état. Le Pilotage montre ce qui
 * demande une décision. Aucune ne répond à la question qu'on se pose en
 * s'asseyant — et ces réponses existaient, éparpillées sur cinq pages.
 *
 * UN SEUL APPEL RÉSEAU : `/api/apercu` agrège tout côté serveur. Six requêtes
 * depuis un téléphone en 4G, c'est un écran qui se remplit par morceaux pendant
 * deux secondes, sur la page qu'on ouvre justement pour un coup d'œil.
 *
 * CE QUI EST DÉLIBÉRÉMENT ABSENT : aucune valeur approchée. Un champ que le
 * serveur rend `null` s'affiche « — », jamais une estimation plausible : on
 * cesserait de vérifier ailleurs.
 */
(function () {
  "use strict";

  const J = window.Crush, el = J.el;

  /* La devise vient de l'API, elle n'est pas supposee ici. Le tracker compte en
   * DOLLARS (`cost_usd`) : afficher « € » etait un montant juste dans la
   * mauvaise monnaie, ce qui est plus trompeur qu'un champ vide. */
  function argent(v, devise) {
    if (v == null) return "—";
    const symbole = devise === "EUR" ? " €" : " $";
    return Number(v).toLocaleString("fr-FR", {
      minimumFractionDigits: 2, maximumFractionDigits: 2,
    }) + symbole;
  }

  function mo(octets) {
    if (octets == null) return "—";
    return (octets / 1048576).toFixed(1) + " Mo";
  }

  function dans(iso) {
    if (!iso) return "";
    const cible = new Date(iso), delta = (cible - Date.now()) / 1000;
    if (delta < 0) return "passé";
    if (delta < 3600) return "dans " + Math.round(delta / 60) + " min";
    if (delta < 86400) return "dans " + Math.round(delta / 3600) + " h";
    return "dans " + Math.round(delta / 86400) + " j";
  }

  /* Une aire remplie, pas le filet de `Crush.sparkline` : sur une tuile sombre,
   * une ligne d'un pixel disparait, et une depense se lit a son VOLUME plus
   * qu'a sa pente. Sept valeurs, donc pas de lissage : chaque jour reste un
   * sommet identifiable. */
  function courbe(serie, teinte) {
    const l = 214, h = 40, ns = "http://www.w3.org/2000/svg";
    const max = Math.max.apply(null, serie) || 1;
    const pts = serie.map(function (v, i) {
      return [(i / (serie.length - 1)) * l, h - (v / max) * (h - 6) - 3];
    });
    const svg = document.createElementNS(ns, "svg");
    svg.setAttribute("viewBox", "0 0 " + l + " " + h);
    svg.setAttribute("width", "100%");
    svg.setAttribute("height", String(h));
    svg.setAttribute("preserveAspectRatio", "none");

    const trace = pts.map(function (p, i) {
      return (i ? "L" : "M") + p[0].toFixed(1) + " " + p[1].toFixed(1);
    }).join(" ");

    const aire = document.createElementNS(ns, "path");
    aire.setAttribute("d", trace + " L" + l + " " + h + " L0 " + h + " Z");
    aire.setAttribute("fill", teinte);
    aire.setAttribute("fill-opacity", ".13");
    svg.appendChild(aire);

    const ligne = document.createElementNS(ns, "path");
    ligne.setAttribute("d", trace);
    ligne.setAttribute("fill", "none");
    ligne.setAttribute("stroke", teinte);
    ligne.setAttribute("stroke-width", "1.6");
    ligne.setAttribute("stroke-linejoin", "round");
    svg.appendChild(ligne);

    // Le dernier point marque : c'est celui qu'on cherche, aujourd'hui.
    const dernier = pts[pts.length - 1];
    const point = document.createElementNS(ns, "circle");
    point.setAttribute("cx", dernier[0].toFixed(1));
    point.setAttribute("cy", dernier[1].toFixed(1));
    point.setAttribute("r", "2.6");
    point.setAttribute("fill", teinte);
    svg.appendChild(point);
    return svg;
  }

  /* ── Briques ──────────────────────────────────────────────────────────────── */

  function tuile(etiquette, valeur, lignes, accent, teinte) {
    const t = el("div", { class: "a-tuile" });
    /* Les memes couleurs que les types du graphe : les deux vues doivent
     * parler la meme langue, sinon chacune a son code et on les relit. */
    if (teinte) t.style.setProperty("--a-teinte", teinte);
    t.appendChild(el("div", { class: "a-tuile-lbl", text: etiquette }));
    const v = el("div", { class: "a-tuile-val", text: valeur });
    if (accent) v.style.color = accent;
    t.appendChild(v);
    (lignes || []).forEach(function (l) {
      if (l == null || l === "") return;
      t.appendChild(el("div", { class: "a-tuile-sub", text: l }));
    });
    return t;
  }

  function section(titre, compte, contenu) {
    const s = el("div", { class: "a-sec" });
    const h = el("div", { class: "a-sec-hd" });
    h.appendChild(el("span", { class: "a-sec-t", text: titre }));
    if (compte) h.appendChild(el("span", { class: "a-sec-n", text: compte }));
    s.appendChild(h);
    s.appendChild(contenu);
    return s;
  }

  function pastilles(liste) {
    const bloc = el("div", { class: "a-pastilles" });
    liste.forEach(function (r) {
      const p = el("span", { class: "a-pastille" });
      p.dataset.actif = String(!!r.actif);
      p.appendChild(el("span", { class: "a-point" }));
      p.appendChild(el("span", { text: r.nom }));
      if (r.detail) p.title = r.detail;
      bloc.appendChild(p);
    });
    return bloc;
  }

  /* ── Rendu ────────────────────────────────────────────────────────────────── */

  async function render(hote) {
    hote.innerHTML = "";
    hote.appendChild(el("div", { class: "a-charge", text: "Relevé en cours…" }));

    let d;
    try {
      d = await J.api.get("/api/apercu");
    } catch (e) {
      hote.innerHTML = "";
      hote.appendChild(el("div", { class: "a-vide", text: "Relevé impossible : " + e.message }));
      return;
    }

    hote.innerHTML = "";
    const page = el("div", { class: "a-page" });

    /* En-tête */
    const tete = el("div", { class: "a-tete" });
    const gauche = el("div");
    gauche.appendChild(el("h1", { class: "a-titre", text: d.assistant }));
    gauche.appendChild(el("div", {
      class: "a-sous",
      text: (d.presence && d.presence.resume) || "état de joignabilité inconnu",
    }));
    tete.appendChild(gauche);
    tete.appendChild(el("div", {
      class: "a-heure",
      text: new Date(d.horodatage).toLocaleTimeString("fr-FR", { hour: "2-digit", minute: "2-digit" }),
    }));
    page.appendChild(tete);

    /* Trois tuiles */
    const tuiles = el("div", { class: "a-tuiles" });
    tuiles.appendChild(tuile(
      "Cerveau",
      d.cerveau.backend,
      [d.cerveau.modele, d.cerveau.reflexion ? "réflexion active" : "réflexion inactive",
       "voix · " + d.cerveau.voix],
      null,
      "#a06ee0",
    ));

    const memLignes = [
      mo(d.memoire.octets),
      d.memoire.sauvegarde_lisible === "inconnu"
        ? "aucune sauvegarde"
        : "sauvegarde " + d.memoire.sauvegarde_lisible,
      d.memoire.archives ? d.memoire.archives + " archive(s)" : "",
    ];
    tuiles.appendChild(tuile(
      "Mémoire",
      d.memoire.faits == null ? "—" : d.memoire.faits + " faits",
      memLignes,
      d.memoire.sauvegarde_inquiete ? "#d99a3c" : null,
      d.memoire.sauvegarde_inquiete ? "#d99a3c" : "#4a90e2",
    ));

    const coutTuile = tuile(
      "Coût",
      argent(d.cout.aujourd_hui, d.cout.devise),
      ["aujourd'hui", d.cout.mois != null ? argent(d.cout.mois, d.cout.devise) + " ce mois" : ""],
      null,
      "#d4af6a",
    );
    if (d.cout.serie && d.cout.serie.length > 1) {
      const enveloppe = el("div", { class: "a-spark" });
      enveloppe.appendChild(courbe(d.cout.serie, "#d4af6a"));
      const bornes = el("div", { class: "a-spark-lbl" });
      bornes.appendChild(el("span", { text: "7 derniers jours" }));
      bornes.appendChild(el("span", {
        text: "max " + argent(Math.max.apply(null, d.cout.serie), d.cout.devise),
      }));
      enveloppe.appendChild(bornes);
      coutTuile.appendChild(enveloppe);
    }
    tuiles.appendChild(coutTuile);
    page.appendChild(tuiles);

    /* À traiter */
    const init = el("div", { class: "a-liste" });
    if (d.initiatives.en_attente) {
      d.initiatives.liste.forEach(function (i) {
        const l = el("a", { class: "a-ligne", href: "/command" });
        l.appendChild(el("span", { class: "a-fleche", text: "▸" }));
        l.appendChild(el("span", { class: "a-ligne-t", text: i.titre }));
        const marque = el("span", { class: "a-marque" });
        marque.dataset.chaud = String(i.priorite === "high");
        marque.textContent = i.decision ? "décision" : i.priorite;
        l.appendChild(marque);
        init.appendChild(l);
      });
      if (d.initiatives.en_attente > d.initiatives.liste.length) {
        init.appendChild(el("div", {
          class: "a-plus",
          text: "et " + (d.initiatives.en_attente - d.initiatives.liste.length) + " autre(s)",
        }));
      }
    } else {
      init.appendChild(el("div", {
        class: "a-vide",
        text: d.initiatives.en_attente === null
          ? "file des initiatives illisible"
          : "rien qui demande ton attention",
      }));
    }
    const colonnes = el("div", { class: "a-colonnes" });
    const colG = el("div", { class: "a-col" });
    const colD = el("div", { class: "a-col" });
    colonnes.appendChild(colG);
    colonnes.appendChild(colD);
    page.appendChild(colonnes);

    colG.appendChild(section(
      "À traiter",
      d.initiatives.en_attente
        ? d.initiatives.en_attente + (d.initiatives.haute ? " · " + d.initiatives.haute + " urgente(s)" : "")
        : "",
      init,
    ));

    /* À regarder — vient de l'Écosystème, pas d'un second comptage : deux
       chiffres qui divergent enverraient chercher une panne au mauvais endroit. */
    const eco = d.ecosysteme || {};
    const aRegarder = el("div", { class: "a-liste" });
    (eco.a_regarder || []).forEach(function (m) {
      const l = el("a", { class: "a-ligne", href: "/capabilities#ecosysteme" });
      l.appendChild(el("span", { class: "a-etat", text: m.etat === "degrade" ? "!" : "○" }));
      const corps = el("div", { class: "a-ligne-corps" });
      corps.appendChild(el("div", { class: "a-ligne-t", text: m.nom + " — " + m.detail }));
      if (m.remede) corps.appendChild(el("div", { class: "a-remede", text: "→ " + m.remede }));
      l.appendChild(corps);
      aRegarder.appendChild(l);
    });
    if (!(eco.a_regarder || []).length) {
      aRegarder.appendChild(el("div", { class: "a-vide", text: "tous les maillons tiennent" }));
    }
    colG.appendChild(section(
      "À regarder",
      eco.ok != null ? eco.ok + " ok · " + (eco.degrade || 0) + " dégradé · " + (eco.absent || 0) + " absent" : "",
      aRegarder,
    ));

    /* Relié / dormant */
    const relies = (d.relie || []).filter(function (r) { return r.actif; });
    const dorment = (d.relie || []).filter(function (r) { return !r.actif; });
    const liens = el("div");
    if (relies.length) {
      liens.appendChild(el("div", { class: "a-etiq", text: "relié" }));
      liens.appendChild(pastilles(relies));
    }
    if (dorment.length) {
      liens.appendChild(el("div", { class: "a-etiq", text: "dormant" }));
      liens.appendChild(pastilles(dorment));
    }
    colD.appendChild(section("Ce qui est branché", relies.length + " / " + (d.relie || []).length, liens));

    /* Boucles */
    const boucles = el("div", { class: "a-liste" });
    (d.boucles || []).forEach(function (b) {
      const l = el("div", { class: "a-ligne" });
      l.appendChild(el("span", { class: "a-ligne-t", text: b.nom }));
      l.appendChild(el("span", { class: "a-cadence", text: b.cadence }));
      l.appendChild(el("span", { class: "a-quand", text: dans(b.quand) }));
      if (b.detail) l.title = b.detail;
      boucles.appendChild(l);
    });
    const silence = d.silence || {};
    if (silence.plage) {
      boucles.appendChild(el("div", {
        class: "a-plus",
        text: "Silence " + silence.plage
          + (silence.urgent_passe ? " — les décisions urgentes passent quand même" : " — rien ne part"),
      }));
    }
    colD.appendChild(section("Ce qui tourne tout seul", (d.boucles || []).length + " boucles", boucles));

    hote.appendChild(page);
  }

  window.CrushApercu = { render: render };
})();
